src/main/java/io/cdap/plugin/google/shopping/content/source/ShoppingContentRecordReader.java

```diff
@@ -0,0 +1,60 @@
+package io.cdap.plugin.google.shopping.content.source;
+
+import org.apache.hadoop.io.Text;
+import org.apache.hadoop.mapreduce.InputSplit;
+import org.apache.hadoop.mapreduce.RecordReader;
+import org.apache.hadoop.mapreduce.TaskAttemptContext;
+
+import java.io.IOException;
+import java.util.List;
+
+/**
+ * A {@link RecordReader} to read the {@link ShoppingContentSplit}.
+ */
+public class ShoppingContentRecordReader extends RecordReader<Text, ProductWritable> {
+
+  private List<ProductWritable> productList;
+  private int currentIndex = 0;
+  private ShoppingContentSplit shoppingContentSplit;
+
+  @Override
+  public void initialize(InputSplit inputSplit, TaskAttemptContext taskAttemptContext) {
+    shoppingContentSplit = (ShoppingContentSplit) inputSplit;
+  }
+
+  @Override
+  public boolean nextKeyValue() throws IOException {
```

Comment:
We shouldn't have to store all of the page contents in memory either. Every time nextKeyValue() is called, we should be returning the next Product from the current page, or fetching the next page and returning the first Product from that page.

Reply:
agree, this is the current behavior?

Reply:
this isn't the current behavior, the client is currently storing everything from the page into a List. Instead of iterating through pages of Products, the client should be iterating through Products.